code/renderers/svelte/src/render.ts

```diff
@@ -1,46 +1,82 @@
 import global from 'global';
-
 import type { Store_RenderContext, ArgsStoryFn } from '@storybook/types';
 import type { SvelteComponentTyped } from 'svelte';
 // eslint-disable-next-line import/no-extraneous-dependencies
 import PreviewRender from '@storybook/svelte/templates/PreviewRender.svelte';
 
 import type { SvelteFramework } from './types';
 
-const { document } = global;
+const componentsByDomElementKey = new Map<string, SvelteComponentTyped>();
 
-let previousComponent: SvelteComponentTyped | null = null;
+const STORYBOOK_ROOT_ID = 'storybook-root';
 
-function cleanUpPreviousStory() {
-  if (!previousComponent) {
+function teardown(domElementKey: string) {
+  if (!componentsByDomElementKey.has(domElementKey)) {
     return;
   }
-  previousComponent.$destroy();
-  previousComponent = null;
+  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion -- we know it exists because we just checked
+  componentsByDomElementKey.get(domElementKey)!.$destroy();
+
+  // in docs mode, the element we render to is a child of the element with the story id
+  // TODO: this feels brittle. If we ever change the structure in the <Story /> component, this will break
+  const rootElement =
+    domElementKey === STORYBOOK_ROOT_ID
+      ? global.document.getElementById(domElementKey)
+      : global.document.getElementById(domElementKey).children[0];
+  rootElement.innerHTML = '';
+  componentsByDomElementKey.delete(domElementKey);
 }
 
 export function renderToDOM(
-  { storyFn, kind, name, showMain, showError, storyContext }: Store_RenderContext<SvelteFramework>,
+  {
+    storyFn,
+    kind,
+    name,
+    id,
+    showMain,
+    showError,
+    storyContext,
+    forceRemount,
+  }: Store_RenderContext<SvelteFramework>,
   domElement: Element
 ) {
-  cleanUpPreviousStory();
+  // in docs mode we're rendering multiple stories to the DOM, so we need to key by the story id
+  const domElementKey = storyContext.viewMode === 'docs' ? id : STORYBOOK_ROOT_ID;
 
-  const target = domElement || document.getElementById('storybook-root');
+  const existingComponent = componentsByDomElementKey.get(domElementKey);
```

Comment:
Would it not be simpler/more reliable to maintain a map `domElement` -> `component`? I suspect that'd resolve the issue above also.

Reply:
you're right, I changed it and that works now with the improved teardown mechanism.